### PR TITLE
feat: add proxy composable

### PR DESCRIPTION
**Pull request by nekosaur**
## Description
Composable can be used to easily set up two-way sync of prop, while keeping internal value state if prop is not defined. If prop is defined it will always be the source of truth.

## Motivation and Context
This attempts to solve some issues we've had when trying to make our components work both with and without setting a v-model. Sometimes internal values have been overriding values set through the prop.

## How Has This Been Tested?
unit tests and playground

## Markup:

<details>

```vue
<template>
  <div>
    <foo
      text="value prop defined, no change handler"
      :value="foo"
    />
    <foo
      text="value prop defined, with change handler"
      :value="bar"
      @update:value="v => bar = v"
    />
    <foo
      text="no value prop defined, no change handler"
    />
    <foo
      text="no value prop defined, change handler"
      @update:value="v => log(`no value prop defined, change handler triggered with ${v}`)"
    />
    <button @click="() => defined = !defined">switch if defined, currently {{ defined ? 'yes' : 'no' }}</button>
    <foo v-bind="props" />
  </div>
</template>

<script lang="ts">
  import { h, defineComponent, ref, computed } from 'vue'
  import { useProxiedModel } from '../src/composables'

  const Foo = defineComponent({
    props: {
      text: String,
      value: {
        type: String,
        default: 'x',
      },
    },
    setup (props, context) {
      const syncedValue = useProxiedModel(props, context, 'value', '')

      return () => h('div', {
        onClick: () => {
          syncedValue.value = syncedValue.value === 'x' ? ' ' : 'x'
        },
      }, [props.text, ' - ', '[', syncedValue.value, ']'])
    },
  })

  export default defineComponent({
    components: {
      Foo,
    },
    setup () {
      const foo = ref('x')
      const bar = ref('x')
      const asd = ref('y')

      const defined = ref(false)

      const props = computed(() => {
        const obj: Record<string, any> = {
          text: 'value prop dynamically defined',
          'onUpdate:value': (v: any) => console.log(v),
        }
        if (defined.value) {
          obj.value = asd.value
        }
        return obj
      })

      return {
        foo,
        bar,
        defined,
        props,
        log: v => console.log(v),
      }
    },
  })
</script>

```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [ ] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

File: packages/vuetify/src/util/helpers.ts [hunk 1]
```diff
@@ -146,7 +146,7 @@ export function convertToUnit (str: string | number | null | undefined, unit = '
 }
 
 export function kebabCase (str: string): string {
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/util/helpers.ts:148]: `string | null | undefined` or remove `(str || '')` and null checking in tests

File: packages/vuetify/src/composables/proxiedModel.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import { ref, computed, getCurrentInstance, SetupContext, Ref } from 'vue'
```
Comment (johnleider) [inline packages/vuetify/src/composables/proxiedModel.ts:1]:
```suggestion
// Utilities
import { ref, computed, getCurrentInstance, SetupContext, Ref } from 'vue'
```

File: packages/vuetify/src/util/__tests__/helpers.spec.ts [hunk 3]
```diff
@@ -1,4 +1,4 @@
-import Vue from 'vue/dist/vue.common.js'
+// import Vue from 'vue/dist/vue.common.js'
```
Comment (johnleider) [inline packages/vuetify/src/util/__tests__/helpers.spec.ts:1]:
```suggestion
```

File: packages/vuetify/src/util/__tests__/helpers.spec.ts [hunk 4]
```diff
@@ -9,6 +9,7 @@ import {
   getObjectValueByPath,
   humanReadableFileSize,
   sortItems,
+  kebabCase,
```
Comment (johnleider) [inline packages/vuetify/src/util/__tests__/helpers.spec.ts:12]: move up one line, alpha ordering 😠

File: packages/vuetify/src/util/__tests__/helpers.spec.ts [hunk 5]
```diff
@@ -362,3 +363,19 @@ describe('helpers', () => {
     expect(items).toStrictEqual([{ string: 'bar', number: 3 }, { string: 'baz', number: 2 }, { string: 'baz', number: 1 }, { string: 'foo', number: 1 }])
   })
 })
+
+describe('kebabCase', () => {
+  it('should convert string to kebab case', () => {
```
Comment (johnleider) [inline packages/vuetify/src/util/__tests__/helpers.spec.ts:368]: should use it.each

File: packages/vuetify/src/util/__tests__/helpers.spec.ts [hunk 6]
```diff
@@ -362,3 +363,19 @@ describe('helpers', () => {
...
+    expect(kebabCase('PascalCase')).toBe('pascal-case')
+    expect(kebabCase('camelCase')).toBe('camel-case')
+    expect(kebabCase('kebab-case')).toBe('kebab-case')
+    expect(kebabCase('ALLCAPS')).toBe('allcaps')
+    expect(kebabCase('this is a sentence')).toBe('this-is-a-sentence')
+    expect(kebabCase('WhatIsThis')).toBe('what-is-this')
+  })
+
+  it('should handle null values', () => {
+    expect(kebabCase(null)).toBe('')
```
Comment (johnleider) [inline packages/vuetify/src/util/__tests__/helpers.spec.ts:378]: should use it.each

**Review by johnleider — CHANGES_REQUESTED** [review]

File: packages/vuetify/src/composables/proxiedModel.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  const internal = ref(transformIn(propIsDefined.value ? props[prop] : defaultValue)) as Ref<Inner>
+
+  return computed<Inner>({
+    get () {
+      if (propIsDefined.value) return transformIn(props[prop])
+      else return internal.value
+    },
+    set (newValue) {
+      internal.value = newValue
+      context.emit(`update:${prop}`, transformOut(newValue))
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/composables/proxiedModel.ts:34]: `vm.emit()`?

File: packages/vuetify/src/util/__tests__/helpers.spec.ts [hunk 8]
```diff
@@ -362,3 +363,19 @@ describe('helpers', () => {
...
 })
+
+describe('kebabCase', () => {
+  it('should convert string to kebab case', () => {
+    expect(kebabCase('PascalCase')).toBe('pascal-case')
+    expect(kebabCase('camelCase')).toBe('camel-case')
+    expect(kebabCase('kebab-case')).toBe('kebab-case')
+    expect(kebabCase('ALLCAPS')).toBe('allcaps')
+    expect(kebabCase('this is a sentence')).toBe('this-is-a-sentence')
+    expect(kebabCase('WhatIsThis')).toBe('what-is-this')
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/util/__tests__/helpers.spec.ts:374]: this test looks like repeated PascalCase test

**Review by jacekkarczmarczyk — APPROVED** [review]
i'd remove unrelated changes from PR and unnecessary !! in computed, but otherwise looks good

**Review by KaelWD — APPROVED** [review]

**Pull request merged**